extension/api.js

```diff
@@ -0,0 +1,189 @@
+          name: "addonsSearchExperiment.onSearchEngineModified",
+          register: (fire) => {
+            const onSearchEngineModifiedObserver = (
+              aSubject,
+              aTopic,
+              aData
+            ) => {
+              if (
+                aTopic !== SEARCH_TOPIC_ENGINE_MODIFIED ||
+                // We are only interested in these modified types.
+                !["engine-added", "engine-removed"].includes(aData)
+              ) {
+                return;
+              }
+
+              searchInitialized.then(() => {
+                fire.async();
+              });
+            };
+
+            Services.obs.addObserver(
+              onSearchEngineModifiedObserver,
+              SEARCH_TOPIC_ENGINE_MODIFIED
+            );
+
+            return () => {
+              Services.obs.removeObserver(
+                onSearchEngineModifiedObserver,
+                SEARCH_TOPIC_ENGINE_MODIFIED
+              );
+            };
+          },
+        }).api(),
+
+        // `onBeforeRedirect` is an event very similar to the public web
+        // request `onBeforeRedirect` but it is registered in the privileged
+        // code to make sure that we can read properties on a valid channel
+        // (wrapper). The registered listeners will received the `addonId` in
+        // addition to these props: `requestId`, `url` and `redirectUrl`.
+        onBeforeRedirect: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onBeforeRedirect",
+          register: (fire, filter) => {
+            const listener = ({ requestId, url, redirectUrl }) => {
+              const wrapper = ChannelWrapper.getRegisteredChannel(
+                requestId,
+                context.extension.policy,
+                context.xulBrowser.frameLoader.remoteTab
+              );
+
+              // When we detect a redirect, we read the request property,
+              // hoping to find an add-on ID corresponding to the add-on that
+              // initiated the redirect. It might not return anything when the
+              // redirect is a search server-side redirect but it can also be
+              // caused by an error.
+              const addonId = wrapper?.channel
+                ?.QueryInterface(Ci.nsIPropertyBag)
+                ?.getProperty("redirectedByExtension");
+
+              fire.async({ addonId, redirectUrl, requestId, url });
+            };
+
+            // See: toolkit/components/extensions/parent/ext-webRequest.js
+            let filter2 = {};
+            if (filter.urls) {
+              let perms = new MatchPatternSet([
+                ...extension.allowedOrigins.patterns,
+                ...extension.optionalOrigins.patterns,
+              ]);
+
+              filter2.urls = ExtensionUtils.parseMatchPatterns(filter.urls);
+            }
+
+            WebRequest.onBeforeRedirect.addListener(
+              listener,
+              filter2,
```

Comment:
The filter should also include `types: ["main_frame"]` since you're only interested in top-level navigation requests.